crates/polars-core/src/frame/row/mod.rs

```diff
@@ -83,40 +83,47 @@ pub fn coerce_data_type<A: Borrow<DataType>>(datatypes: &[A]) -> DataType {
     try_get_supertype(lhs, rhs).unwrap_or(String)
 }
 
-/// Infer schema from rows and set the supertypes of the columns as column data type.
+/// Infer the schema of rows by determining the supertype of the values.
+///
+/// Field names are set as `column_0`, `column_1`, and so on.
 pub fn rows_to_schema_supertypes(
     rows: &[Row],
     infer_schema_length: Option<usize>,
 ) -> PolarsResult<Schema> {
+    let dtypes = rows_to_supertypes(rows, infer_schema_length)?;
+    let schema = dtypes_to_schema(dtypes);
+    Ok(schema)
+}
+
+/// Infer the schema data types of rows by determining the supertype of the values.
+pub fn rows_to_supertypes(
+    rows: &[Row],
+    infer_schema_length: Option<usize>,
+) -> PolarsResult<Vec<DataType>> {
     polars_ensure!(!rows.is_empty(), NoData: "no rows, cannot infer schema");
 
-    // no of rows to use to infer dtype
-    let max_infer = infer_schema_length.unwrap_or(rows.len());
-    let mut dtypes: Vec<PlIndexSet<DataType>> = vec![PlIndexSet::new(); rows[0].0.len()];
+    let max_infer = infer_schema_length.unwrap_or_else(|| rows.len());
 
+    let mut dtypes: Vec<PlIndexSet<DataType>> = vec![PlIndexSet::new(); rows[0].0.len()];
     for row in rows.iter().take(max_infer) {
-        for (val, types_set) in row.0.iter().zip(dtypes.iter_mut()) {
-            types_set.insert(val.into());
+        for (val, dtypes_set) in row.0.iter().zip(dtypes.iter_mut()) {
+            dtypes_set.insert(val.into());
         }
     }
 
     dtypes
         .into_iter()
-        .enumerate()
-        .map(|(i, types_set)| {
-            let dtype = if types_set.is_empty() {
-                DataType::Unknown
-            } else {
-                dtypes_to_supertype(&types_set)?
-            };
-            Ok(Field::new(format!("column_{i}").as_ref(), dtype))
-        })
-        .collect::<PolarsResult<_>>()
+        .map(|dtypes_set| dtypes_to_supertype(&dtypes_set))
+        .collect()
 }
 
 /// Infer schema from rows and set the first no null type as column data type.
-pub fn rows_to_schema_first_non_null(rows: &[Row], infer_schema_length: Option<usize>) -> Schema {
-    // no of rows to use to infer dtype
+pub fn rows_to_schema_first_non_null(
+    rows: &[Row],
+    infer_schema_length: Option<usize>,
+) -> PolarsResult<Schema> {
+    polars_ensure!(!rows.is_empty(), NoData: "no rows, cannot infer schema");
```

Comment:
I added this ensure to avoid a panic. This makes it a breaking change on the Rust side, as it changes the return type.